### PR TITLE
Reducing complexity of voting power function

### PR DESCRIPTION
**Pull request by RonTuretzky**
#84

### CODE REVIEW

File: src/YieldDistributor.sol [hunk 1]
```diff
@@ -140,32 +140,18 @@ contract YieldDistributor is IYieldDistributor, OwnableUpgradeable {
...
-        } while (_currentCheckpoint._key > _end);
-
-        /// Initialize voting power with the latest checkpoint thats within the interval (or nearest to it)
-        uint48 _latestKey = _currentCheckpoint._key < _start ? uint48(_start) : _currentCheckpoint._key;
-        uint256 _totalVotingPower = _currentCheckpoint._value * (_end - _latestKey);
-
-        if (_latestKey == _start) return _totalVotingPower;
+        uint256 _totalVotingPower;
 
         for (uint32 i = _currentCheckpointIndex; i > 0;) {
```
Comment (bagelface) [inline src/YieldDistributor.sol]: with this refactor moving the logic into a single for loop, I think the variable `_currentCheckpointIndex` would be better renamed to `_numCheckpoints` and then when setting us this for loop you can instead do:

```
for (uint32 i = _numCheckpoints - 1; i >= 0; --i)` {
  _currentCheckpoint = _sourceContract.checkpoints(_account, i);
  ...
}
```

File: src/YieldDistributor.sol [hunk 2]
```diff
@@ -140,32 +140,18 @@ contract YieldDistributor is IYieldDistributor, OwnableUpgradeable {
...
-            /// Latest checkpoint voting power is calculated when initializing `_totalVotingPower`, so we pre-decrement the index here
             _currentCheckpoint = _sourceContract.checkpoints(_account, --i);
+            if (_currentCheckpoint._key <= _end) {
+                uint48 effectiveStart = _currentCheckpoint._key < _start ? uint48(_start) : _currentCheckpoint._key;
```
Comment (bagelface) [inline src/YieldDistributor.sol]: prefix `_` for internal variable

File: src/YieldDistributor.sol [hunk 3]
```diff
@@ -140,32 +140,18 @@ contract YieldDistributor is IYieldDistributor, OwnableUpgradeable {
...
-        uint48 _latestKey = _currentCheckpoint._key < _start ? uint48(_start) : _currentCheckpoint._key;
-        uint256 _totalVotingPower = _currentCheckpoint._value * (_end - _latestKey);
-
-        if (_latestKey == _start) return _totalVotingPower;
+        uint256 _totalVotingPower;
 
         for (uint32 i = _currentCheckpointIndex; i > 0;) {
-            /// Latest checkpoint voting power is calculated when initializing `_totalVotingPower`, so we pre-decrement the index here
             _currentCheckpoint = _sourceContract.checkpoints(_account, --i);
+            if (_currentCheckpoint._key <= _end) {
```
Comment (bagelface) [inline src/YieldDistributor.sol]: line break before big if block for readability

File: src/YieldDistributor.sol [hunk 4]
```diff
@@ -140,32 +140,18 @@ contract YieldDistributor is IYieldDistributor, OwnableUpgradeable {
...
+                uint48 effectiveStart = _currentCheckpoint._key < _start ? uint48(_start) : _currentCheckpoint._key;
+                _totalVotingPower += _currentCheckpoint._value * (_end - effectiveStart);
 
-            /// Add voting power for the sub-interval to the total
-            _totalVotingPower += _currentCheckpoint._value * (_latestKey - _currentCheckpoint._key);
+                if (effectiveStart == _start) break;
```
Comment (bagelface) [inline src/YieldDistributor.sol:151]: can prob remove this line break and combine these two since they're both single line control flow statements

**Review by bagelface — APPROVED** [review]
a few suggested changes, but logic changes seem fine

File: src/YieldDistributor.sol (hunk 1 above)
Comment (RonTuretzky) [reply src/YieldDistributor.sol]: addressed with 9f92e21

File: src/YieldDistributor.sol (hunk 2 above)
Comment (RonTuretzky) [reply src/YieldDistributor.sol]: addressed with b6c1b77

File: src/YieldDistributor.sol (hunk 3 above)
Comment (RonTuretzky) [reply src/YieldDistributor.sol]: addressed with 51931c6

File: src/YieldDistributor.sol (hunk 4 above)
Comment (RonTuretzky) [reply src/YieldDistributor.sol:151]: addressed with d87945a

**Review by ShaNagbaImuru — APPROVED** [review]
LGTM, thanks!

**Pull request merged**